fastestimator/estimator.py

```diff
@@ -64,92 +64,104 @@ def __init__(self,
+                self.network.unload_epoch()
+
+    def _check_keys(self):
+        for mode in self.pipeline.get_modes():
+            pipeline_all_outputs = self.pipeline.get_all_output_keys(mode, self.epochs)
+            network_all_outputs = self.network.get_all_output_keys(mode, self.epochs)
+            assert self.trace_inputs[mode].issubset(pipeline_all_outputs | network_all_outputs), "found missing key"
+            self.network.effective_inputs[mode] = self.network.get_effective_input_keys(mode, self.epochs)
+            self.network.effective_outputs[mode] = network_all_outputs.intersection(self.trace_inputs[mode])
 
     def _prepare_system(self):
-        self.system = System(mode="train",
-                             global_step=0,
-                             batch_size=None,
-                             num_devices=get_num_devices(),
-                             log_steps=self.log_steps,
-                             total_epochs=self.epochs,
-                             total_steps=None,
-                             epoch_idx=0,
-                             batch_idx=0)
+        system = System(mode="train",
+                        global_step=0,
+                        num_devices=get_num_devices(),
+                        log_steps=self.log_steps,
+                        total_epochs=self.epochs,
+                        epoch_idx=0,
+                        batch_idx=0)
+        system.network = self.network
         for trace in self.traces:
-            trace.system = self.system
+            trace.system = system
+        return system
 
     def _prepare_traces(self):
-        loss_keys = self._initialize_loss_keys()
-        if self.traces is None:
-            self.traces = []
-        self.traces = to_list(self.traces)
-        self.monitor_names = set(filter(None, to_list(self.monitor_names))).union(loss_keys)
-        self.traces.insert(0, TrainEssential(monitor_names=self.monitor_names))
-        if self.do_eval:
-            self.traces.append(EvalEssential(loss_keys=loss_keys))
-        for trace in self.traces:
-            self.trace_inputs = self.trace_inputs.union(set(filter(None, to_list(trace.inputs))))
-            self.monitor_names = self.monitor_names.union(set(filter(None, to_list(trace.log_names))))
-        self.traces.append(Logger(log_names=self.monitor_names, loss_names=loss_keys))
-
-    def _initialize_loss_keys(self) -> Set[str]:
-        loss_keys = set()
-        for op in self.network.ops:
-            if isinstance(op, UpdateOp):
-                loss_keys = loss_keys.union(set(to_list(op.inputs)))
-        return loss_keys
+        self.trace_inputs = dict()
+        traces = [trace for trace in self.traces]
+        loss_keys = self.network.get_loss_keys()
+        monitor_names = set(filter(None, to_list(self.monitor_names))).union(loss_keys)
+        traces.insert(0, TrainEssential(monitor_names=monitor_names))
+        modes = self.pipeline.get_modes() - {"test"}
```

Comment:
why are we removing 'test' from modes?

Reply:
because `Trace` is not designed for test or inference.

Reply:
XAI traces would want to run there still

Reply:
Inferencing should not go through Estimator, therefore not through trace.

Reply:
That seems like we're shooting ourselves in our foot. traces might be useful

Reply:
We will think about it when we implement the inferencing API :)